### ISSUE TITLE
GPTQ Qwen3.5-35B checkpoint produces incoherent BPE-fragment output on real hardware (correctness bug, not sampling)

### ISSUE THREAD

**Issue by aangelinsf** [Performant-Labs/FreeToken-Intel#147]
## Summary

End-to-end real-checkpoint generation against `Qwen/Qwen3.5-35B-A3B-GPTQ-Int4` (the checkpoint driving epic #134) completes without crashing or OOMing, but produces incoherent output: a small set of repeating BPE subword fragments across mixed scripts, never resolving into real words in either the `<think>` reasoning block or an eventual answer.

This was validated with the checkpoint's own recommended sampling settings (`do_sample=true, temperature=1.0, top_k=20, top_p=0.95`, per its `generation_config.json`), ruling out the earlier suspicion that greedy/temperature=0 decoding was the cause (that produced a single repeated token — a different, more obviously-degenerate failure mode, also ruled out separately).

## Evidence

Prompt: `In one short sentence, what is the capital of France?` (Qwen3.5 chat template, `<think>\n` continuation, 22 prompt tokens).

Config: `moe_backend="offload"`, `dtype=bfloat16`, `device=cpu`, `max_seq_len_override=320`, `num_page_override=320`, `page_size=1`, `max_running_req=1`. `torch.manual_seed(0)`.

200 tokens generated, `generate done in 303.1s (1.52s/tok)`, `peak RSS 21.61 GiB` -- no crash, no OOM, no new GPU exec-queue resets.

Generated text (truncated):
```
大戏endahantro vestiarosendah casinarosendahendahirlines prestiixo prestiditorantroendah
intentsantro但最终퀴antro퀴endah-ENDarosantroarosixoendaharos但最终calar퀴 preziendahcalar
prezi而代之 prestiendahantro-ENDendaharos퀴大戏而代之 preziantro-END偽ixo-END偽arosieta prezi
casinarosendah preziendahditorendahixoixoaros prestiirlinesixoixoaros prestiieta preziendaharos...
```

A small, recurring set of subword fragments (`endah`, `aros`, `presti`, `-END`, `ixo`, `antro`, mixed CJK) repeats throughout, rather than degrading gracefully into plausible-but-wrong words. This pattern -- not "stuck reasoning" or "wrong topic," but a bounded vocabulary of nonsense fragments -- is consistent with a numerical correctness bug feeding bad logits, not a prompting/sampling issue.

## What's already ruled out

- **Sampling/decoding config**: checkpoint's own recommended sampling settings were used; a separate greedy-decoding run produced a distinctly different failure (single repeated token), confirming decoding strategy is not the root cause of *this* failure.
- **Offload/cache machinery itself**: a fast diagnostic reusing `tests/test_qwen35_gptq_e2e_loader.py`'s fabricated hybrid-checkpoint fixtures (real GPTQ per-expert tensors, including `.g_idx`) drove a real multi-step `Engine.generate()` against a tiny synthetic GPTQ checkpoint and produced varied, non-degenerate tokens -- suggesting the packed-bank LRU/slot-accessor/dequant-at-compute machinery from epic #134 (issues #135-138) is not itself the source of this bug.
- **Gross memory/process health**: no OOM, no swap thrashing severe enough to corrupt data, no crash.

## Suspected areas (not yet isolated)

- GPTQ dequant numerical correctness at real scale (group_size=128, g_idx=`k // group_size`, zero-point `+1` convention) -- the synthetic-fixture diagnostic uses much smaller dimensions and may not exercise all group-boundary/zero-point edge cases the real checkpoint does.
- Gated-DeltaNet hybrid linear-attention state handling (30/40 layers use GDN, 10/40 use full attention every 4th layer) under this offload configuration -- state carried across the real 40-layer depth may behave differently than in short synthetic tests.
- Possible interaction between the dense (non-quantized) weight path and the packed-bank offload path for routed vs. shared experts.

## Suggested next steps

- Compare a dequantized GPTQ expert tensor slice against a reference dequant (e.g. AutoGPTQ's own dequant, or a manual numpy reimplementation) for a real layer/expert from this checkpoint, to isolate the group_size/g_idx/zero-point math.
- Try `moe_backend` other than `"offload"` (e.g. dense placement) against the same real checkpoint to see whether the bug is offload-specific or present in the base dequant/forward path generally.
- Try the same prompt against a real HF/transformers reference run of this checkpoint (CPU or another accelerator) to establish a known-good baseline for comparison.

## Scope note

This is a **new correctness issue**, distinct from epic #134 (issues #135-138, which validated the cache-budget/packed-bank *loading and memory* mechanics and are already closed). Filing separately per user direction rather than continuing to debug inline.

Co-Authored-By: Claude Sonnet 5 <noreply@anthropic.com>
Claude-Session: https://claude.ai/code/session_01EQVhsUeDoaGusbYPuEkdve

**Comment by aangelinsf** [comment]
Found and fixed one real bug (PR #148): the dequant kernel's zero-point convention was wrong for this checkpoint's format (legacy AutoGPTQ v1's `+1` correction, assumed from documentation, never checked against real bytes -- this checkpoint's GPTQModel-produced `qzeros` store the true zero-point directly). Verified quantitatively against the real checkpoint's tensors.

Re-ran the real 200-token generation with the fix applied: output is still incoherent, but the failure signature changed character (diverse multilingual garbage instead of a small repeating fragment set) -- consistent with a real, partial fix, not a no-op. At least one more correctness bug remains somewhere in the pipeline. Leaving this open; next suspects per the original issue body: GDN hybrid-attention state handling, or another dequant edge case not exercised by the small synthetic fixtures.

Co-Authored-By: Claude Sonnet 5 <noreply@anthropic.com>
Claude-Session: https://claude.ai/code/session_01EQVhsUeDoaGusbYPuEkdve

**Comment by aangelinsf** [comment]
Further investigation since the zero-point fix (PR #148): ruled out several more suspects via careful static review + direct verification against the real checkpoint's bytes, all clean:

- **MoE routing/gating** (`_Qwen35MoE.forward`): top-k=8 (matches the checkpoint's `num_experts_per_tok: 8`), softmax-then-topk-then-renormalize, shared-expert sigmoid gate added separately -- standard Qwen3-MoE architecture, no bug found.
- **Gated-DeltaNet hybrid attention**: the delta-rule recurrence (decay-then-correct), RoPE convention, L2-norm on q/k, and per-request state carry all check out against the reference formula.
- **Offload-cache LRU slot bookkeeping**: traced `materialize_layer`/`ensure_experts`/`copy_missing`/`_resync_slot_for_id` -- the eviction invariant holds, `SlotWeightAccessor` resolves gate_up and down from one consistent slot snapshot per step, no staleness across layers/steps. (Found one unrelated, harmless dead-code reference: `offload_cache.py`'s `__all__` names a `GPTQ_INT4_BANK_NAMES` that doesn't exist -- nothing imports via `*` so it's inert, but worth a cleanup.)
- **GPTQ dequant at real scale, both projections**: directly dequantized real layer-0/expert-0 `gate_proj` AND `down_proj` tensors from the checkpoint with the fixed kernel -- both produce the expected ~0 mean (symmetric-calibrated weights), shapes match the documented bank contract. `moe_intermediate_size=512` divides evenly by `group_size=128` (4 groups), so the group-boundary edge case doesn't apply here.
- Confirmed `embed_tokens`/`lm_head`/attention weights are plain (unquantized) in the real checkpoint, matching the loader's assumptions.

Nothing pins down the remaining bug yet. Given every individually-testable piece of the pipeline checks out clean in isolation, the next real diagnostic likely requires either (a) an actual layer-by-layer intermediate-activation comparison against a reference implementation of this checkpoint (e.g. a real transformers/vLLM load, if one is feasible on available hardware) to find which layer's output first diverges, or (b) suspecting precision loss compounding across 40 layers of bf16 dequant-at-compute rather than a single localized bug. Both are bigger undertakings than the audits so far.

Co-Authored-By: Claude Sonnet 5 <noreply@anthropic.com>
Claude-Session: https://claude.ai/code/session_01EQVhsUeDoaGusbYPuEkdve

**Comment by aangelinsf** [comment]
**Correction:** PR #148 (the "zero-point fix") was wrong -- reverting it in #149.

I installed \`gptqmodel\` (the actively-maintained, authoritative GPTQ reference implementation) and ran its own dequant utility directly against this checkpoint's real config and tensors. It determines this checkpoint uses AutoGPTQ v1 qzeros (\`checkpoint_format\` is absent from \`quantization_config\`, and gptqmodel's own \`_uses_gptq_v1_qzeros\` treats a missing field as v1) and applies the \`+1\` correction -- exactly reproducing the ORIGINAL (pre-#148) code's output. #148's justification (near-zero dequantized weight mean without the \`+1\`) was a real but misleading signal: calibrated GPTQ weights are not reliably zero-mean, so that heuristic was the wrong correctness test. My apologies for the churn -- should have checked against an authoritative reference from the start instead of reasoning from a plausibility heuristic.

Net effect: the GPTQ dequant convention was correct all along. Issue #147's actual bug (incoherent generation) is still open and still unexplained. Extensive line-by-line comparison of routing, GDN attention, RMSNorm (both variants), the router formula, embedding, and both dequant projections against the real HF reference source (\`transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py\`, which happens to already be vendored in this project's venvs) all match exactly -- no divergence found anywhere checked so far.

Next productive step is likely either (a) an actual reference-model forward pass (now that gptqmodel + the real HF modeling code are available, a full CPU load/forward comparison is plausible, if slow, given this is a 35B model) to find which layer's output first diverges from a trusted implementation, or (b) continuing targeted source-level comparison into areas not yet checked (the full-attention softmax-attention kernel/paged-KV read-back itself, or the causal-conv/state-carry mechanics under this project's specific paged/offload setup vs. the reference's cache_params object).

Co-Authored-By: Claude Sonnet 5 <noreply@anthropic.com>
Claude-Session: https://claude.ai/code/session_01EQVhsUeDoaGusbYPuEkdve

**Comment by aangelinsf** [comment]
Found and fixed another real, source-verified bug (PR #150): \`_GatedDeltaNet._conv\` never applied the model's SiLU activation to the causal-conv output before splitting into q/k/v -- the real HF reference (\`causal_conv1d_fn\`/\`causal_conv1d_update\`, both called with \`activation=self.activation\`) does. Found by directly diffing against the vendored real reference source (this project's venvs happen to already carry \`transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py\`), not a heuristic -- confirmed correct, unlike the zero-point mistake in #148/#149.

This affects every linear-attention (GDN) layer -- 30 of 40 -- every token. Also found and fixed the same gap in \`tests/reference_qwen35.py\`'s "independent" reference, which had copied the port's buggy behavior instead of the true upstream source (explaining why no test caught this).

Re-ran the real 200-token generation with the fix applied: still incoherent, though the token stream changed slightly (2 of the first 20 sampled tokens differ from the pre-fix run) -- confirms a real numerical effect, just not sufficient alone. At least one more bug remains.

Given the extent of investigation so far (routing, RMSNorm both variants, router formula, embedding, RoPE, output gating, LRU cache, GPTQ dequant, and now the GDN conv activation -- all directly diffed against the real reference source), the one major area not yet compared line-by-line against the reference is the full-attention path's actual softmax-attention computation and paged-KV read-back (10 of 40 layers) -- the projections/RoPE/output-gate around it were checked, but not the attention kernel itself or how it interacts with this project's paged KV cache. That's the next place to look.

Co-Authored-By: Claude Sonnet 5 <noreply@anthropic.com>
Claude-Session: https://claude.ai/code/session_01EQVhsUeDoaGusbYPuEkdve

**Comment by aangelinsf** [comment]
Low-memory follow-up verification (no real checkpoint needed): built three independent, numerically exact comparisons of our port's per-layer forward math against the real vendored HF reference implementation (\`transformers/models/qwen3_5_moe/modeling_qwen3_5_moe.py\`) at tiny synthetic scale -- not the hand-written \`tests/reference_qwen35.py\`, which turned out to share our port's own bug, but the actual upstream reference classes, with identical weights loaded into both:

- **Gated-Delta-Net** (\`_GatedDeltaNet\` vs \`Qwen3_5MoeGatedDeltaNet\`): exact match (max diff ~1e-6, float32 noise) -- confirms the SiLU fix (#150) is correct and complete.
- **Full attention** (\`_Qwen35Attention\` vs \`Qwen3_5MoeAttention\`): exact match (max diff ~1.5e-8) -- projections, partial RoPE, gating, GQA repeat, causal softmax all verified bit-correct.
- **MoE routing + experts + shared expert** (\`_Qwen35MoE\` vs \`Qwen3_5MoeSparseMoeBlock\`): exact match (diff == 0.0) -- router softmax/topk/renormalize, gate_up fusion order, shared-expert sigmoid gate all verified bit-correct.

Also traced the offload-specific compute glue (\`_forward_offload_core\`/\`SlotWeightAccessor\`): it calls the same \`_expert_compute\` formula already verified above. The code has an existing, acknowledged comment that the offload path's accumulation order differs slightly from the in-RAM path's, producing ~1e-7 ULP differences that get amplified by the GDN recurrence's chaotic dynamics over many decode steps (referencing a prior issue #18) -- real floating-point non-associativity, not a correctness bug, and it doesn't explain incoherence from the very first generated token.

At this point every individually-testable piece of the architecture is proven correct against ground truth. Given that, the two live hypotheses are: (a) a bug specific to conditions only present at real scale (256 experts / 40 layers / real LRU eviction pressure / the real GPTQ checkpoint's actual weight values) that no synthetic test reproduces, or (b) the model is genuinely unusually numerically sensitive in a way that isn't a "bug" to fix in this port. Distinguishing between these needs either more memory headroom for a full reference forward pass on the real checkpoint, or further real-hardware runs with more targeted instrumentation (e.g., dumping per-layer stats from a REAL reference run using the same real weights, not just tiny synthetic ones).

Co-Authored-By: Claude Sonnet 5 <noreply@anthropic.com>
Claude-Session: https://claude.ai/code/session_01EQVhsUeDoaGusbYPuEkdve

**Comment by aangelinsf** [comment]
Comprehensive low-memory validation, now complete. Built a real, working genuine end-to-end reference harness (not the hand-written \`tests/reference_qwen35.py\`, which shared our port's own bug) using two things: the real HF \`transformers\` classes already vendored in this project's venvs, and \`tiny-random/qwen3.5-moe\` -- a real, tiny, publicly-hosted HuggingFace test checkpoint of this exact architecture (found while looking for a low-memory way to test at real scale; downloading it surfaced and let me fix a real, separate bug, see below).

**Results, all EXACT matches (bit-for-bit or float32-noise-level) against the true reference implementation:**
1. Single-layer isolated components (tiny synthetic weights): Gated-Delta-Net, full attention, MoE routing+experts+shared-expert -- each individually exact.
2. **Full multi-layer forward pass** (real 4-layer/128-expert \`tiny-random/qwen3.5-moe\` checkpoint, identical weights loaded into both our port and the real HF model): logits match to float32 noise (diff ~3.6e-7).
3. **Full 8-step greedy autoregressive decode** (same checkpoint): token-for-token EXACT match, including Gated-Delta-Net recurrent state carry and KV-cache append across real multiple steps.
4. **Forced real LRU eviction** (cache_size=20 slots for 128 experts, so eviction happens every single step): still an exact match.
5. **GPTQ + offload + forced eviction + multi-step decode, all combined** (a small fabricated GPTQ checkpoint, 16 experts, cache_size=6 so real eviction happens constantly, 6 real decode steps): still an EXACT match.

At this point every mechanism in the pipeline -- individually and in every combination I could construct -- is proven bit-correct against real ground truth. This is about as thorough as validation gets without the real 35B checkpoint itself.

**Bonus: found and fixed a real, separate bug** while setting this up (PR #151, open): \`_expert_source_names\` generated only an un-indexed key spelling (\`experts.gate_up_proj\`) that no real checkpoint file actually uses (real files always store one tensor per expert, \`experts.{e}.gate_proj.weight\`). This was invisible on every GPTQ checkpoint tested so far because the GPTQ path uses a different, correct classification method -- it only affects plain (non-GPTQ) Qwen3.5-MoE checkpoints, which is why loading \`tiny-random/qwen3.5-moe\` (non-quantized) failed until fixed.

**Given this, the most likely remaining explanations for #147's incoherence are things only reproducible at real scale/real weights:**
- The real checkpoint's actual trained weight VALUES (not random ones) triggering a numerically-adjacent edge case my random-weight tests never hit (e.g. a specific magnitude/precision interaction).
- Something scale-dependent (256 real experts / 40 real layers vs. my max of 16 experts / 2 layers) that no code path I've read suggests should matter, but hasn't been ruled out by direct test.
- The real checkpoint's actual \`mrope_interleaved\`/\`mrope_section\` position handling for a REAL multi-token prompt, if HF's real generation pipeline computes position_ids differently than the naive \`arange\` this session's tests all used (worth one more targeted check).

Next step, when there's RAM headroom for the real checkpoint again: a real per-layer hidden-state dump (already have the diagnostic script) compared statistically isn't enough, given everything else now checks out -- what's actually needed is either (a) getting a working GPTQ CPU backend (gptqmodel + optimum) to run a real reference forward on the ACTUAL checkpoint for direct tensor comparison, or (b) checking the real position_ids construction for a multi-token prompt specifically.

Co-Authored-By: Claude Sonnet 5 <noreply@anthropic.com>
Claude-Session: https://claude.ai/code/session_01EQVhsUeDoaGusbYPuEkdve
